base/iterators.jl

```diff
@@ -648,62 +603,85 @@ julia> collect(Iterators.product(1:2,3:5))
+indices(P::ProductIterator) = _prod_indices(P.iterators)
+_prod_indices(::Tuple{}) = ()
+_prod_indices(t::Tuple) = tuple(_prod_indices1(t[1], iteratorsize(t[1]))..., _prod_indices(tail(t))...)
+_prod_indices1(a, ::HasShape)  = indices(a)
+_prod_indices1(a, ::HasLength) = (OneTo(length(a)),)
+_prod_indices1(a, A) =
+    throw(ArgumentError("Cannot compute indices for object of type $(typeof(a))"))
 
-    nv2 = st[3]
-    if isnull(nv2)
-        v2, s2 = next(p.b, s2)
-    else
-        v2 = nv2.value
-    end
+ndims(p::ProductIterator) = length(indices(p))
+length(P::ProductIterator) = prod(size(P))
+_length(p::ProductIterator) = prod(map(unsafe_length, indices(p)))
 
-    if done(p.a, s1)
-        return (v1,v2), (start(p.a), s2, Nullable{eltype(nv2)}(), done(p.b,s2))
-    end
-    return (v1,v2), (s1, s2, Nullable(v2), false)
+iteratoreltype(::Type{ProductIterator{Tuple{}}}) = HasEltype()
+iteratoreltype(::Type{ProductIterator{Tuple{I}}}) where {I} = iteratoreltype(I)
+function iteratoreltype(::Type{ProductIterator{T}}) where {T<:Tuple}
+    I = tuple_type_head(T)
+    P = ProductIterator{tuple_type_tail(T)}
+    iteratoreltype(I) == EltypeUnknown() ? EltypeUnknown() : iteratoreltype(P)
 end
 
-@inline next(p::Prod2, st) = prod_next(p, st)
-@inline done(p::AbstractProdIterator, st) = st[4]
+Base.eltype(P::ProductIterator) = _prod_eltype(P.iterators)
+_prod_eltype(::Tuple{}) = Tuple{}
+_prod_eltype(t::Tuple) = Base.tuple_type_cons(eltype(t[1]),_prod_eltype(tail(t)))
 
-# n iterators
-struct Prod{I1, I2<:AbstractProdIterator} <: AbstractProdIterator
-    a::I1
-    b::I2
+function Base.start(P::ProductIterator)
+    iterators = P.iterators
+    states = map(start, iterators)
+    if any(map(done, iterators, states))
+        return ProductIteratorState{typeof(states),eltype(P)}(true, states)
+    else
+        return ProductIteratorState{typeof(states),eltype(P)}(false, _prod_start(iterators, states)...)
+    end
 end
-product(a, b, c...) = Prod(a, product(b, c...))
-
-eltype(::Type{Prod{I1,I2}}) where {I1,I2} = tuple_type_cons(eltype(I1), eltype(I2))
-
-iteratoreltype(::Type{Prod{I1,I2}}) where {I1,I2} = and_iteratoreltype(iteratoreltype(I1),iteratoreltype(I2))
-iteratorsize(::Type{Prod{I1,I2}}) where {I1,I2} = prod_iteratorsize(iteratorsize(I1),iteratorsize(I2))
+Base.next(P::ProductIterator, state) = state.values, ProductIteratorState(_prod_next(P.iterators, state.states, state.values)...)
+Base.done(P::ProductIterator, state) = state.done
 
-@inline function next(p::Prod, st)
-    x = prod_next(p, st)
-    ((x[1][1],x[1][2]...), x[2])
+_prod_start(iterators::Tuple{}, states::Tuple{}) = (), ()
+function _prod_start(iterators, states)
+    val, state = next(iterators[1], states[1])
+    tailstates, tailvals = _prod_start(tail(iterators), tail(states))
+    return (state, tailstates...), (val, tailvals...)
+end
+_prod_next(iterators::Tuple{}, states::Tuple{}, values::Tuple{}) = true, (), ()
+function _prod_next(iterators, states, values)
```

Comment:
Oh, duh, it's definitely not helping here because this function is not named `next`. This is pretty good evidence we could take that bonus out and use the "pure" algorithm.